File: mobile/android/app/src/main/kotlin/com/opensourceortho/lite/ScanImport.kt
package com.opensourceortho.lite

/** Offline scanner geometry import. Material paths and external textures are never opened. */
object ScanImport {
    private val whitespace = Regex("\\s+")
    val extensions = setOf("stl", "obj", "ply", "asc", "xyz", "pts")
    const val formats = "STL, OBJ, PLY, ASC, XYZ, PTS"
    fun supports(filename: String) = filename.substringAfterLast('.', "").lowercase() in extensions

    fun decode(bytes: ByteArray, filename: String, check: () -> Unit = {}): StlMesh {
        require(bytes.size <= StlMesh.MAX_BYTES) { "Scan exceeds mobile preview limit." }
        return when (filename.substringAfterLast('.', "").lowercase()) {
            "stl" -> StlMesh.decode(bytes, check)
            "obj" -> textMesh(bytes, true, false, check)
            "ply" -> PlyImport.decode(bytes, check)
            "asc", "xyz" -> textMesh(bytes, false, false, check)
            "pts" -> textMesh(bytes, false, true, check)
            else -> error("Unsupported scan. Export $formats from your scanner.")
        }
    }

    internal fun validateTextLines(bytes: ByteArray, offset: Int = 0, check: () -> Unit = {}) {
        var length = 0
        for (i in offset until bytes.size) {
            if (i % 262144 == 0) check()
            val b = bytes[i].toInt()
            length = if (b == 10 || b == 13) 0 else length + 1
            require(b != 0 && length <= 16384) { "Invalid or overlong scan text line." }
        }
    }

    private fun textMesh(bytes: ByteArray, obj: Boolean, counted: Boolean, check: () -> Unit): StlMesh {
        validateTextLines(bytes, 0, check)
        val points = FloatValues()
        val triangles = FloatValues()
        var expected: Int? = null
        bytes.inputStream().bufferedReader().useLines { lines ->
            lines.forEachIndexed { i, line ->
                if (i % 1024 == 0) check()
                val clean = line.substringBefore('#').trim()
                if (clean.isEmpty()) return@forEachIndexed
                val tokens = clean.split(whitespace)
                if (counted && expected == null) {
                    require(tokens.size == 1) { "PTS must begin with a point count." }
                    expected = tokens[0].toInt()
                    require(expected!! in 1..StlMesh.MAX_TRIANGLES) { "Invalid point count." }
                } else if (obj && tokens[0] == "f") {
                    val indices = tokens.drop(1).map {
                        val raw = it.substringBefore('/').toInt()
                        require(raw != 0) { "OBJ indices start at one." }
                        if (raw > 0) raw - 1 else points.size / 3 + raw
                    }
                    appendFace(indices, points, triangles)
                } else if (!obj || tokens[0] == "v") {
                    val coordinates = if (obj) tokens.drop(1) else tokens
                    require(coordinates.size >= 3 && (!obj || coordinates.size in listOf(3,4,6)) &&
                        (!obj || coordinates.size != 4 || coordinates[3].toFloatOrNull() == 1f)) { "Incomplete point." }
                    coordinates.take(3).forEach {
                        val v = it.toFloat()
                        require(v.isFinite() && kotlin.math.abs(v) <= 1_000_000f) { "Invalid coordinates." }
                        points.add(v)
                    }
                    require(points.size / 3 <= StlMesh.MAX_TRIANGLES) { "Too many points." }
                } else if (tokens[0] in listOf("l", "curv", "curv2", "surf")) {
                    error("Export a polygon mesh or point cloud from your scanner.")
                }
            }
        }
        require(expected == null || expected == points.size / 3) { "PTS point count does not match its data." }
        return StlMesh.build((if (triangles.isEmpty()) points else triangles).toFloatArray(), check, triangles.isEmpty())
    }

    internal fun appendFace(indices: List<Int>, points: List<Float>, triangles: FloatValues) {
        require(indices.size in 3..64 && indices.all { it >= 0 && it < points.size / 3 } && indices.distinct().size == indices.size) { "Invalid polygon indices." }
        require(triangles.size / 9 + indices.size - 2 <= StlMesh.MAX_TRIANGLES) { "Too many triangles." }
        if (indices.size > 3) validatePolygon(indices, points)
        for (i in 1 until indices.size - 1) {
            for (index in listOf(indices[0], indices[i], indices[i+1])) {
                for (axis in 0..2) triangles.add(points[index*3+axis])
            }
        }
    }

    private fun validatePolygon(indices: List<Int>, points: List<Float>) {
        val p = indices.map { index -> DoubleArray(3) { points[index*3+it].toDouble() } }
        fun cross(a: DoubleArray, b: DoubleArray) = doubleArrayOf(a[1]*b[2]-a[2]*b[1],a[2]*b[0]-a[0]*b[2],a[0]*b[1]-a[1]*b[0])
        val n = cross(DoubleArray(3) { p[1][it]-p[0][it] }, DoubleArray(3) { p[2][it]-p[1][it] })
        val length = kotlin.math.sqrt(n.sumOf { it*it })
        require(length > 0) { "Degenerate polygon. Export triangulated geometry." }
        for (i in p.indices) {
            val distance = (0..2).sumOf { (p[i][it]-p[0][it])*n[it] } / length
            require(kotlin.math.abs(distance) <= 0.0001) { "Non-planar face. Export triangulated geometry." }
            val edge = DoubleArray(3) { p[(i+1)%p.size][it]-p[i][it] }
            for (j in p.indices) {
                if (j == i || j == (i+1)%p.size) continue
                val side = cross(edge, DoubleArray(3) { p[j][it]-p[i][it] })
                require((0..2).sumOf { side[it]*n[it] } > 0) {
                    "Concave or self-intersecting face. Export triangulated geometry."
                }
            }
        }
    }
}

File: mobile/ios/Sources/OpenSourceOrthoLiteKit/ScanImport.swift
import Foundation

/// Scanner exports stay on-device. Texture paths are never opened or fetched.
public enum ScanImport {
    public static let extensions = ["stl", "obj", "ply", "asc", "xyz", "pts"]
    public static let formats = "STL, OBJ, PLY, ASC, XYZ, PTS"
    public static func supports(_ filename: String) -> Bool {
        extensions.contains(URL(fileURLWithPath: filename).pathExtension.lowercased())
    }

    public static func decode(_ data: Data, filename: String) throws -> STLMesh {
        guard data.count <= STLMesh.maxBytes else { throw STLMesh.LoadError.tooLarge }
        switch URL(fileURLWithPath: filename).pathExtension.lowercased() {
        case "stl": return try STLMesh.decode(data)
        case "obj": return try textMesh(data, obj: true, counted: false)
        case "ply": return try PLYImport.decode(data)
        case "asc", "xyz": return try textMesh(data, obj: false, counted: false)
        case "pts": return try textMesh(data, obj: false, counted: true)
        default: throw STLMesh.LoadError.invalid
        }
    }

    private static func textMesh(_ data: Data, obj: Bool, counted: Bool) throws -> STLMesh {
        try validateTextLines(data)
        guard let text = String(data: data, encoding: .utf8) else { throw STLMesh.LoadError.invalid }
        var points: [Float] = [], triangles: [Float] = []
        var expected: Int?
        for (lineIndex, line) in text.split(whereSeparator: \.isNewline).enumerated() {
            if lineIndex % 1024 == 0 { try Task.checkCancellation() }
            let tokens = line.split(separator: "#", maxSplits: 1, omittingEmptySubsequences: false)[0]
                .split(whereSeparator: \.isWhitespace)
            if tokens.isEmpty { continue }
            if counted && expected == nil {
                guard tokens.count == 1, let count = Int(tokens[0]), count > 0,
                      count <= STLMesh.maxTriangles else { throw STLMesh.LoadError.invalid }
                expected = count; continue
            }
            if obj && tokens[0] == "f" {
                let indices = try tokens.dropFirst().map { token -> Int in
                    guard let raw = Int(token.split(separator: "/", omittingEmptySubsequences: false)[0]), raw != 0 else {
                        throw STLMesh.LoadError.invalid
                    }
                    return raw > 0 ? raw - 1 : points.count / 3 + raw
                }
                try appendFace(indices, points: points, triangles: &triangles)
            } else if !obj || tokens[0] == "v" {
                let coordinates = obj ? Array(tokens.dropFirst()) : Array(tokens)
                guard coordinates.count >= 3,
                      !obj || [3,4,6].contains(coordinates.count),
                      !obj || coordinates.count != 4 || Float(coordinates[3]) == 1 else { throw STLMesh.LoadError.invalid }
                for token in coordinates.prefix(3) {
                    guard let value = Float(token), value.isFinite, abs(value) <= 1_000_000 else {
                        throw STLMesh.LoadError.invalid
                    }
                    points.append(value)
                }
                guard points.count / 3 <= STLMesh.maxTriangles else { throw STLMesh.LoadError.tooLarge }
            } else if ["l", "curv", "curv2", "surf"].contains(String(tokens[0])) {
                throw STLMesh.LoadError.invalid
            }
        }
        if let expected, expected != points.count / 3 { throw STLMesh.LoadError.invalid }
        return try STLMesh.build(triangles.isEmpty ? points : triangles, pointCloud: triangles.isEmpty)
    }

    /// Triangle and convex planar polygon exports; reject ambiguous concave/non-planar faces.
    static func appendFace(_ indices: [Int], points: [Float], triangles: inout [Float]) throws {
        guard (3...64).contains(indices.count), indices.allSatisfy({ $0 >= 0 && $0 < points.count / 3 }),
              Set(indices).count == indices.count else { throw STLMesh.LoadError.invalid }
        guard triangles.count / 9 + indices.count - 2 <= STLMesh.maxTriangles else { throw STLMesh.LoadError.tooLarge }
        if indices.count > 3 { try validatePolygon(indices, points: points) }
        for i in 1..<(indices.count - 1) {
            for index in [indices[0], indices[i], indices[i+1]] {
                triangles.append(contentsOf: points[index*3..<index*3+3])
            }
        }
    }

    private static func validatePolygon(_ indices: [Int], points: [Float]) throws {
        let p = indices.map { i in (0..<3).map { Double(points[i*3+$0]) } }
        func cross(_ a: [Double], _ b: [Double]) -> [Double] {
            [a[1]*b[2]-a[2]*b[1], a[2]*b[0]-a[0]*b[2], a[0]*b[1]-a[1]*b[0]]
        }
        let a = (0..<3).map { p[1][$0]-p[0][$0] }, b = (0..<3).map { p[2][$0]-p[1][$0] }
        let n = cross(a,b), length = sqrt(n.reduce(0) { $0+$1*$1 })
        guard length > 0 else { throw STLMesh.LoadError.invalid }
        for i in p.indices {
            let d = (0..<3).map { p[i][$0]-p[0][$0] }
            let distance = zip(d,n).reduce(0) { $0+$1.0*$1.1 } / length
            let u = (0..<3).map { p[(i+1)%p.count][$0]-p[i][$0] }
            guard abs(distance) <= 0.0001 else { throw STLMesh.LoadError.invalid }
            // Every other vertex must lie strictly on the same side of this edge.
            // Consecutive turns alone incorrectly accept self-intersecting stars.
            for j in p.indices where j != i && j != (i+1)%p.count {
                let v = (0..<3).map { p[j][$0]-p[i][$0] }
                let side = zip(cross(u,v),n).reduce(0) { $0+$1.0*$1.1 }
                guard side > 0 else { throw STLMesh.LoadError.invalid }
            }
        }
    }
}
